docs/ffi/ffi.md

```diff
@@ -0,0 +1,467 @@
+```
+
+Starting at the bottom, we're using the XLA-provided macro `XLA_FFI_DEFINE_HANDLER` to generate some boilerplate which will expand into a function called `RmsNorm` with the appropriate signature.
+But, the important stuff here is all in the call to `ffi::Ffi::Bind()`, where we define the input and output types, and the types of any parameters.
+
+Then, in `RmsNormImpl`, we accept `ffi::Buffer` arguments which include information about the buffer shape, and pointers to the underlying data.
+In this implementation, we treat all leading dimensions of the buffer as batch dimensions, and perform RMS normalization over the last axis.
+`GetDims` is a helper function providing support for this batching behavior.
+
+### Building and registering an FFI handler
+
+Now that we have our minimal FFI wrapper implemented, we need to expose this function (`RmsNorm`) to Python.
+In this tutorial, we use [nanobind](https://nanobind.readthedocs.io) to define a tiny Python module, but it is also possible to compile `RmsNorm` into a shared library and load it using [ctypes](https://docs.python.org/3/library/ctypes.html), a pattern which we discuss below.
+
+For this example, our nanobind module is defined as follows:
+
+```c++
+#include <type_traits>
+
+#include "nanobind/nanobind.h"
+#include "xla/ffi/api/c_api.h"
+
+namespace nb = nanobind;
+
+template <typename T>
+nb::capsule EncapsulateFfiCall(T *fn) {
+  static_assert(std::is_invocable_r_v<XLA_FFI_Error *, T, XLA_FFI_CallFrame *>,
+                "Encapsulated function must be and XLA FFI handler");
+  return nb::capsule(reinterpret_cast<void *>(fn), "xla._CUSTOM_CALL_TARGET");
+}
+
+NB_MODULE(rms_norm, m) {
+  m.def("rms_norm", []() { return EncapsulateFfiCall(RmsNorm); });
+}
+```
+
+With this in place, we can now compile our library.
+Here we use CMake, but you should be able to use your favorite build system without too much trouble.
+The full `CMakeLists.txt` can be downloaded [here](CMakeLists.txt).
+
+```{code-cell} ipython3
+:tags: [hide-output]
+
+!cmake -DCMAKE_BUILD_TYPE=Release -B _build .
+!cmake --build _build
+!cmake --install _build
+```
+
+With this compiled library in hand, we now need to register this handler with XLA via the {func}`~jax.lib.xla_client.register_custom_call_target` function.
+In our nanobind module above, we implemented a Python function called `rms_norm` which returns a [`PyCapsule`](https://docs.python.org/3/c-api/capsule.html) containing a function pointer to the C++ function `RmsNorm`.
+This "capsule" is the object that we pass to `register_custom_call_target`:
+
+```{code-cell} ipython3
+from jax.lib import xla_client
+import rms_norm as rms_norm_lib
+
+xla_client.register_custom_call_target(
+  "rms_norm", rms_norm_lib.rms_norm(), platform="cpu", api_version=1
+)
+```
+
+We note that the `api_version=1` keyword in the call to `register_custom_call_target` is important because it indicates that our function pointer implements this FFI interface.
```

Comment:
s/this/typed/?

Alternatively, we can say that `api_version=0` is the legacy FFI interface which we are not covering in the doc?